### PR TITLE
Fix clippy warnings

### PR DESCRIPTION
**Pull request by rukai**
- [x] Tested on all platforms changed - Should be no change in functionality.
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

There are still some lints left, but this should be all the non-controversial ones. The rest can be ignored or fixed in a separate PR.

I decided to just ignore new_without_default as it seems really aggressive to me. Happy to just remove the ignore and leave the decision for a follow up PR.

Make sure to review the diff with whitespace disabled to keep the match changes readable.

### CODE REVIEW

**Review by madsmtm — CHANGES_REQUESTED** [review]
Related: https://github.com/rust-windowing/winit/pull/2076.

Quick notes: Agree with most of these, maybe we should enable `clippy` in CI in the future?

However, I think we should discuss the `match` changes in the examples: I think there's value in keeping them as-is, because it provides a nice "template" that users can easily copy-paste and extend on. Same goes for `winit` developers, I think it's useful to be able to easily drop in an extra pattern when debugging an example.

**Comment by rukai** [comment]
I think the match changes make sense but happy to roll them back to keep this PR easy to merge.
Which of these change should I rollback:
* combining 2 match statements into 1
* replacing `match` with `if let`
?

**Comment by madsmtm** [comment]
Honestly, can I choose both? At first glance I meant only the `match -> if let`, but I'm unsure if e.g. `examples/mouse_wheel.rs` got any clearer by combining the `match` statements, so I'd like to discuss that as well / at least have it explicitly changed.

**Comment by rukai** [comment]
yep, both was definitely an option!
All the match changes should be reverted now

File: src/platform_impl/linux/x11/util/icon.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ impl Icon {
         data.push(rgba_icon.height as Cardinal);
         let pixels = rgba_icon.rgba.as_ptr() as *const Pixel;
         for pixel_index in 0..pixel_count {
-            let pixel = unsafe { &*pixels.offset(pixel_index as isize) };
+            let pixel = unsafe { &*pixels.add(pixel_index) };
```
Comment (madsmtm) [inline src/platform_impl/linux/x11/util/icon.rs:30]: This one scared me for a second (thought it was suddenly trying to use integer addition, which is strictly not the same as pointer offset), but `pointer::add` is just a convenience for this exact case

**Review by madsmtm — APPROVED** [review]
Looked through it all, it looks fine, nice work!

Would it be possible for you to do the same for the other platforms as well?

File: src/lib.rs [hunk 2]
```diff
@@ -131,6 +131,7 @@
 
 #![deny(rust_2018_idioms)]
 #![deny(rustdoc::broken_intra_doc_links)]
+#![allow(clippy::new_without_default)]
```
[the remark this answers is not in the document]
Comment (rukai) [reply src/lib.rs]: Alright, reverted this so each individual case can be investigated.

**Comment by rukai** [comment]
I do actually have access to macos and windows so im happy to do those as well, but lets land this PR first.

**Review by kchibisov — CHANGES_REQUESTED** [review]
Thx for the PR. It would be nice to finally resolve this issue https://github.com/rust-windowing/winit/issues/1402.

To test platforms other than your host I could recommend you doing something like `cargo +nightly clippy  --target=x86_64-pc-windows-msvc --all-targets ` (this works for like every platform winit supports).

**Comment by rukai** [comment]
If you are happy with the fixes in this PR please land them first.

File: src/platform_impl/linux/x11/ime/callbacks.rs [hunk 3]
```diff
@@ -136,13 +137,17 @@ pub unsafe extern "C" fn xim_instantiate_callback(
...
-            result.expect("Failed to reopen input method");
+        match replace_im(inner) {
+            Ok(()) => {
+                let _ = unset_instantiate_callback(xconn, client_data);
+                (*inner).is_fallback = false;
+            }
+            Err(err) => {
+                if (*inner).is_destroyed {
+                    // We have no usable input methods!
+                    panic!("Failed to reopen input method: {:?}", err);
```
Comment (maroider) [inline src/platform_impl/linux/x11/ime/callbacks.rs:148]: Debug formatting is probably the correct "direct" translation, but is it the correct formatting option here?

File: src/platform_impl/linux/x11/window.rs [hunk 4]
```diff
@@ -431,7 +433,7 @@ impl UnownedWindow {
 
             // Select XInput2 events
             let mask = {
```
Comment (maroider) [inline src/platform_impl/linux/x11/window.rs:435]: Is the extra scope necessary/useful here?

**Review by maroider — CHANGES_REQUESTED** [review]
looks mostly ok to me

File: src/platform_impl/linux/x11/ime/callbacks.rs (hunk 3 above)
Comment (rukai) [reply src/platform_impl/linux/x11/ime/callbacks.rs:148]: yes, I think so.
Not sure what else I could do here.

File: src/platform_impl/linux/x11/util/geometry.rs [hunk 5]
```diff
@@ -248,7 +248,7 @@ impl XConnection {
             );
 
             // The list of children isn't used
-            if children != ptr::null_mut() {
+            if children.is_null() {
```
[the remark this answers is not in the document]
Comment (rukai) [reply src/platform_impl/linux/x11/util/geometry.rs]: A pretty dangerous oversight, yes.
Good catch.

File: src/platform_impl/linux/x11/ime/callbacks.rs (hunk 3 above)
Comment (maroider) [reply src/platform_impl/linux/x11/ime/callbacks.rs:148]: Well, the alternative is "normal" formatting, but it's not that big of a deal either way.

Comment (rukai) [reply src/platform_impl/linux/x11/ime/callbacks.rs:148]: I would just try it if I were at a computer but I'm pretty sure you can't just normal format an enum unless it implements display

File: src/platform_impl/linux/x11/window.rs (hunk 4 above)
Comment (rukai) [reply src/platform_impl/linux/x11/window.rs]: Removed the scope

File: src/platform_impl/linux/x11/ime/callbacks.rs (hunk 3 above)
Comment (maroider) [reply src/platform_impl/linux/x11/ime/callbacks.rs:148]: Ah, alright, that's my bad for not checking what the error type is.

**Review by maroider — APPROVED** [review]
It seems like this should be ready to be merged.

I'm assuming adding clippy to our CI config is intended to be added in a separate PR.

**Pull request merged**